ambry-server/src/test/java/com.github.ambry.server/AmbryRequestsTest.java

```diff
@@ -616,6 +662,87 @@ private Response sendRequestGetResponse(RequestOrResponse request, ServerErrorCo
+   *                          response in the constituents rather than the actual response ({@link GetResponse} and
+   *                          {@link ReplicaMetadataResponse}).
+   * @param forceCheckOpReceived if {@code true}, checks the operation received at the {@link Store} even if
+   *                             there is an error expected. Always checks op received if {@code expectedErrorCode} is
+   *                             {@link ServerErrorCode#No_Error}. Skips the check otherwise.
+   * @throws InterruptedException
+   * @throws IOException
+   */
+  private void sendAndVerifyOperationRequest(RequestOrResponse request, ServerErrorCode expectedErrorCode,
+      Boolean forceCheckOpReceived) throws InterruptedException, IOException {
+    storageManager.resetStore();
+    RequestOrResponseType requestType = request.getRequestType();
+    Response response = sendRequestGetResponse(request,
+        requestType == RequestOrResponseType.GetRequest || requestType == RequestOrResponseType.ReplicaMetadataRequest
+            ? ServerErrorCode.No_Error : expectedErrorCode);
+    if (expectedErrorCode.equals(ServerErrorCode.No_Error) || forceCheckOpReceived) {
+      assertEquals("Operation received at the store not as expected", requestType,
+          MockStorageManager.operationReceived);
+    }
+    if (requestType == RequestOrResponseType.GetRequest) {
+      GetResponse getResponse = (GetResponse) response;
+      for (PartitionResponseInfo info : getResponse.getPartitionResponseInfoList()) {
+        assertEquals("Error code does not match expected", expectedErrorCode, info.getErrorCode());
+      }
+    } else if (requestType == RequestOrResponseType.ReplicaMetadataRequest) {
+      ReplicaMetadataResponse replicaMetadataResponse = (ReplicaMetadataResponse) response;
+      for (ReplicaMetadataResponseInfo info : replicaMetadataResponse.getReplicaMetadataResponseInfoList()) {
+        assertEquals("Error code does not match expected", expectedErrorCode, info.getError());
+      }
+    }
+  }
+
+  /**
+   * @param id the {@link PartitionId} to find the {@link ReplicaId} for on the current node
+   * @return the {@link ReplicaId} corresponding to {@code id} on this data node
+   */
+  private ReplicaId findReplica(PartitionId id) {
+    ReplicaId replicaId = null;
+    for (ReplicaId replica : id.getReplicaIds()) {
+      if (replica.getDataNodeId().equals(dataNodeId)) {
+        replicaId = replica;
+        break;
+      }
+    }
+    return replicaId;
+  }
+
+  /**
+   * Changes the state of the partition
+   * @param id the {@link MockPartitionId} whose status needs to change
+   * @param replicaDown if {@code true}, partition will become
+   * {@link com.github.ambry.clustermap.PartitionState#READ_ONLY}. Otherwise
+   * {@link com.github.ambry.clustermap.PartitionState#READ_WRITE}
+   */
+  private void changePartitionState(PartitionId id, boolean replicaDown) {
```

Comment:
I am a little confused here. If we want to change partition to `READ_ONLY`, we can mark one of the replicas as `Sealed`. This is what `resolvePartitionStatus` method is actually checking with. Why do we mark replica down here?

Reply:
Thanks for catching this ! This code is incorrect (wasn't marking the partition `READ_ONLY`) - fixed now